File: p3.py
#TC= O(n)
#SC = O(1)
# if runs successfully on leetcode: YES


class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        #use split on s
        #s.split = ["dog", "cat", "cat", "dog"]

        s_list = s.split()
        if len(pattern) != len(s_list):
            return False
        p_to_s = {}
        s_to_p = {}
        for i in range(0, len(pattern)):
            if pattern[i] not in p_to_s:
               p_to_s[pattern[i]] = s_list[i] #{a:dog}
            else:
                if p_to_s[pattern[i]] != s_list[i]:
                    return False
            if s_list[i] not in s_to_p:
               s_to_p[s_list[i]] = pattern[i] #{a:dog}
            else:
                if s_to_p[s_list[i]] != pattern[i]:
                    return False
        return True

File: p3.java
// Time Complexity : Time complexity would be O(n) as we have to traverse the string
// Space Complexity : Space complexity would be O(1) as there can be finite number of letters 
// Did this code successfully run on Leetcode :Yes

// Any problem you faced while coding this :
// Had to lookup String.split

// Your code here along with comments explaining your approach
// Using hashmap and hashset. Using hashmap to maintaing correlation between alphabets in string pattern and words in string s. Using hashset to find if there's any case of word getting repeared for different alphabet.

class Solution {
    public boolean wordPattern(String pattern, String s) {
        String[] sarray = s.split(" ");
        if (pattern.length() != sarray.length)
            return false;

        Map<Character, String> map = new HashMap<>();
        Set<String> set = new HashSet<>();

        for (int i = 0; i < pattern.length(); i++) {
            char p = pattern.charAt(i);
            if (!map.containsKey(p)) {
                map.put(p, sarray[i]);
                if (set.contains(sarray[i]))
                    return false;
            }
            if (!map.get(p).equals(sarray[i]))
                return false;

            set.add(sarray[i]);
        }

        return true;
    }
}